zod/app/routes/products.tsx

```diff
@@ -0,0 +1,198 @@
+    (pagination.page - 1) * pagination.size,
+    pagination.page * pagination.size,
+  );
+
+  return json(
+    {
+      products: paginatedProducts,
+      searchParams: {
+        name,
+        minPrice,
+        maxPrice,
+        page: pagination.page,
+        size: pagination.size,
+      },
+      totalPageCount,
+    },
+    {
+      headers: {
+        "Cache-Control": "public, max-age=30, s-maxage=60",
+      },
+    },
+  );
+};
+
+export default function ProductsView() {
+  const loaderData = useLoaderData<typeof loader>();
+  const submit = useSubmit(); // used for select onChange
+  const navigation = useNavigation();
+  const isLoading = navigation.state === "loading";
+
+  return (
+    <div>
+      <h1>Products</h1>
+
+      <Form method="get">
+        {/* Filters */}
+        <label htmlFor="name">Product Name:</label>
+        <input
+          type="text"
+          id="name"
+          name="name"
+          defaultValue={loaderData.searchParams.name}
+        />{" "}
+        <label htmlFor="minPrice">Min Price:</label>
+        <input
+          type="number"
+          id="minPrice"
+          name="minPrice"
+          defaultValue={loaderData.searchParams.minPrice}
+        />{" "}
+        <label htmlFor="maxPrice">Max Price:</label>
+        <input
+          type="number"
+          id="maxPrice"
+          name="maxPrice"
+          defaultValue={loaderData.searchParams.maxPrice}
+        />{" "}
+        <button type="submit" disabled={isLoading}>
+          Search
+        </button>
+        <hr />
+        <ul id="productList">
+          {loaderData.products.length > 0 ? (
+            loaderData.products.map((product) => (
+              <li key={product.id}>
+                <span>
+                  <i>{product.name}</i>{" "}
+                </span>
+                <span>
+                  <b>${product.price}</b>
+                </span>
+              </li>
+            ))
+          ) : (
+            <p>No products found</p>
+          )}
+        </ul>
+        <hr />
+        {/* Pagination */}
+        <span>
+          Page {loaderData.searchParams.page} of {loaderData.totalPageCount}
+        </span>{" "}
+        <button
+          type="submit"
+          name="page"
+          value={loaderData.searchParams.page - 1}
+          disabled={isLoading || loaderData.searchParams.page === 1}
+        >
+          Prev
+        </button>{" "}
+        <button
+          type="submit"
+          name="page"
+          value={loaderData.searchParams.page + 1}
+          disabled={
+            isLoading ||
+            loaderData.searchParams.page === loaderData.totalPageCount
+          }
+        >
+          Next
+        </button>{" "}
+        <label htmlFor="size">Items per Page:</label>
+        <select
+          id="size"
+          name="size"
+          defaultValue={loaderData.searchParams.size}
+          onChange={(event) => {
+            submit(event.currentTarget.form);
+          }}
```

Comment:
Since you're submitting the whole form, this can be done on the `Form` itself

Reply:
Initially, I didn't wanted to implement a filter as you type experience for all the form. But I thought, at least for the items per page select, it would be more intuitive for users to have it submitted on change without manually submitting.

If having it on the whole form will be nicer, maybe I should also add some error info to loader data for all the fields in products route, instead of just throwing.

Reply:
Now that I'm thinking about it, putting the `onChange` on the form, would mean it gets submitted on every change, not only of the selection 😕